### PR TITLE
Integrate mempool with chainstate

### PR DESCRIPTION
**Pull request by royster57**
This PR links the mempool to the chainstate.

### CODE REVIEW

**Comment by TheQuantumPhysicist** [comment]
You may want to update your local copy and rebase/merge. You have lots of duplicated changes.

File: chainstate/src/detail/chainstateref.rs [hunk 1]
```diff
@@ -145,6 +145,10 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
         }
     }
 
+    pub fn make_utxo_view(&self) -> impl UtxosView + '_ {
+        UtxosDB::new(&self.db_tx)
```
[the remark this answers is not in the document]
Comment (royster57) [reply chainstate/src/detail/chainstateref.rs:150]: Done.

File: mempool/src/error.rs [hunk 2]
```diff
@@ -81,6 +82,22 @@ pub enum TxValidationError {
...
     DescendantOfExpiredTransaction,
+    #[error("Chainstate error")]
+    ChainstateError,
+    #[error("Subsystem call error")]
+    CallError,
     #[error("Internal Error.")]
     InternalError,
 }
+
+impl From<subsystem::subsystem::CallError> for TxValidationError {
```
Comment (TheQuantumPhysicist) [inline mempool/src/error.rs:93]: You don't need to do these manually. See how the `#[from]` macro is used in other places to automatically do these for you.

File: mempool/src/pool.rs [hunk 3]
```diff
@@ -174,44 +189,49 @@ impl RollingFeeRate {
...
     ) -> Self {
         Self {
+            chain_config,
             store: MempoolStore::new(),
-            chain_state,
             chainstate_handle,
             max_size: MAX_MEMPOOL_SIZE_BYTES,
             max_tx_age: DEFAULT_MEMPOOL_EXPIRY,
-            rolling_fee_rate: Cell::new(RollingFeeRate::new(clock.get_time())),
+            rolling_fee_rate: parking_lot::RwLock::new(RollingFeeRate::new(clock.get_time())),
```
Comment (TheQuantumPhysicist) [inline mempool/src/pool.rs:234]: I'm not sure we really have to use parking_lot anymore. In a recent version of rust, mutexes have been improved to solve the performance problems that parking_lot solves. Maybe throw in a TODO here to research this some time in the future.

**Review by TheQuantumPhysicist — APPROVED** [review]
Looking good. Please address the few issues I mentioned, then let's merge this.

**Review by TheQuantumPhysicist — APPROVED** [review]
Looking good. Please address the few issues I mentioned, then let's merge this.

File: chainstate/src/interface/chainstate_interface_impl.rs [hunk 4]
```diff
@@ -280,4 +285,49 @@ impl<S: BlockchainStorage> ChainstateInterface for ChainstateInterfaceImpl<S> {
...
+                        }
+                    }
+                } else {
+                    values.push(None)
+                }
+                Ok(values)
+            },
+        );
+
+        eprintln!("time spent getting outpoint {:?}", start.elapsed());
```
[the remark this answers is not in the document]
Comment (royster57) [reply chainstate/src/interface/chainstate_interface_impl.rs:330]: Noted. In any case, this print is no longer needed, so I removed it.

File: chainstate/src/interface/chainstate_interface_impl.rs [hunk 5]
```diff
@@ -280,4 +285,49 @@ impl<S: BlockchainStorage> ChainstateInterface for ChainstateInterfaceImpl<S> {
...
+            .filter(|input| utxo_view.utxo(input.outpoint()).is_some())
+            .cloned()
+            .collect();
+        Ok(available_inputs)
+    }
+
+    fn get_outpoint_values(
+        &self,
+        tx: &Transaction,
+    ) -> Result<Vec<Option<common::primitives::Amount>>, ChainstateError> {
```
[the remark this answers is not in the document]
Comment (royster57) [reply chainstate/src/interface/chainstate_interface_impl.rs:304]: Done.

File: chainstate/src/interface/chainstate_interface_impl.rs [hunk 6]
```diff
@@ -280,4 +285,49 @@ impl<S: BlockchainStorage> ChainstateInterface for ChainstateInterfaceImpl<S> {
...
+        use std::time::Instant;
+        let start = Instant::now();
```
[the remark this answers is not in the document]
Comment (royster57) [reply chainstate/src/interface/chainstate_interface_impl.rs:306]: Noted, will do.

File: chainstate/src/interface/chainstate_interface_impl.rs [hunk 7]
```diff
@@ -280,4 +285,49 @@ impl<S: BlockchainStorage> ChainstateInterface for ChainstateInterfaceImpl<S> {
...
+        let available_inputs = tx
+            .inputs()
+            .iter()
+            .filter(|input| utxo_view.utxo(input.outpoint()).is_some())
+            .cloned()
+            .collect();
+        Ok(available_inputs)
+    }
+
+    fn get_outpoint_values(
```
[the remark this answers is not in the document]
Comment (royster57) [reply chainstate/src/interface/chainstate_interface_impl.rs:301]: Done!

File: mempool/src/pool/store.rs [hunk 8]
```diff
@@ -95,7 +95,8 @@ impl MempoolStore {
 
     // Checks whether the outpoint is to be created by an unconfirmed tx
     pub(super) fn contains_outpoint(&self, outpoint: &OutPoint) -> bool {
-        matches!(self.txs_by_id.get(&outpoint.tx_id().get_tx_id().expect("Not Coinbase").get()),
+        outpoint.tx_id().get_tx_id().is_some()
+            && matches!(self.txs_by_id.get(&outpoint.tx_id().get_tx_id().expect("Not Coinbase").get()),
```
[the remark this answers is not in the document]
Comment (royster57) [reply mempool/src/pool/store.rs:99]: Changed the expect message accordingly.

File: mempool/src/pool/tests.rs [hunk 9]
```diff
@@ -38,396 +52,49 @@ fn dummy_size() {
...
-
-pub(crate) fn create_genesis_tx() -> Transaction {
-    const TOTAL_SUPPLY: u128 = 10_000_000_000_000;
-    let genesis_message = b"".to_vec();
-    let outpoint_source_id = OutPointSourceId::Transaction(Id::new(H256::zero()));
-    let input = TxInput::new(
-        outpoint_source_id,
+#[tokio::test]
+async fn real_size() -> anyhow::Result<()> {
+    let seed = Seed::from_entropy();
```
[the remark this answers is not in the document]
Comment (royster57) [reply mempool/src/pool/tests.rs:57]: Done :)

File: mempool/src/error.rs (hunk 2 above)
Comment (royster57) [reply mempool/src/error.rs:93]: Thanks for reminding me of that!

File: mempool/src/pool.rs (hunk 3 above)
Comment (royster57) [reply mempool/src/pool.rs:234]: Threw in a TODO.

File: chainstate/src/interface/chainstate_interface_impl.rs [hunk 10]
```diff
@@ -280,4 +285,49 @@ impl<S: BlockchainStorage> ChainstateInterface for ChainstateInterfaceImpl<S> {
             .get_token_id_from_issuance_tx(tx_id)
             .map_err(ChainstateError::FailedToReadProperty)
     }
+
+    fn available_inputs(&self, tx: &Transaction) -> Result<Vec<TxInput>, ChainstateError> {
```
[the remark this answers is not in the document]
Comment (royster57) [reply chainstate/src/interface/chainstate_interface_impl.rs]: Agreed. Fixed.

**Pull request merged**